File: solution/backend/mlsys_scheduler/dag.py
"""DAG utilities: topological sort (Kahn's algorithm), adjacency lists, graph I/O identification."""

from __future__ import annotations

from collections import deque

from mlsys_scheduler.models import DAGInfo, Problem


def build_dag(problem: Problem) -> DAGInfo:
    """Analyse the problem DAG and return pre-computed graph info.

    Uses Kahn's algorithm for topological sort (O(V + E)).

    Args:
        problem: The parsed Problem struct.

    Returns:
        DAGInfo with adjacency, topo order, graph inputs/outputs.

    Raises:
        ValueError: If the DAG contains a cycle.
    """
    num_ops = len(problem.ops)
    num_tensors = len(problem.tensors)

    # tensor_producer[tensor_id] = op index that produces it, or -1 for graph inputs
    tensor_producer: dict[int, int] = {}
    # tensor_consumers[tensor_id] = list of op indices that consume it
    tensor_consumers: dict[int, list[int]] = {t: [] for t in range(num_tensors)}

    for op_idx, op in enumerate(problem.ops):
        for tensor_id in op.outputs:
            tensor_producer[tensor_id] = op_idx
        for tensor_id in op.inputs:
            tensor_consumers[tensor_id].append(op_idx)

    # Identify graph inputs (tensors with no producer) and graph outputs (tensors with no consumer)
    graph_inputs: set[int] = set()
    graph_outputs: set[int] = set()
    for tensor_id in range(num_tensors):
        if tensor_id not in tensor_producer:
            tensor_producer[tensor_id] = -1
            graph_inputs.add(tensor_id)
        if not tensor_consumers[tensor_id]:
            graph_outputs.add(tensor_id)

    # Build op-level adjacency: op A -> op B if A produces a tensor consumed by B
    adjacency: dict[int, list[int]] = {i: [] for i in range(num_ops)}
    reverse_adj: dict[int, list[int]] = {i: [] for i in range(num_ops)}

    for op_idx, op in enumerate(problem.ops):
        seen_successors: set[int] = set()
        for tensor_id in op.outputs:
            for consumer_op in tensor_consumers[tensor_id]:
                if consumer_op not in seen_successors:
                    seen_successors.add(consumer_op)
                    adjacency[op_idx].append(consumer_op)
                    reverse_adj[consumer_op].append(op_idx)

    # Kahn's algorithm for topological sort
    in_degree = [len(reverse_adj[i]) for i in range(num_ops)]
    queue: deque[int] = deque(op_idx for op_idx in range(num_ops) if in_degree[op_idx] == 0)
    topo_order: list[int] = []

    while queue:
        op_idx = queue.popleft()
        topo_order.append(op_idx)
        for successor in adjacency[op_idx]:
            in_degree[successor] -= 1
            if in_degree[successor] == 0:
                queue.append(successor)

    if len(topo_order) != num_ops:
        raise ValueError("DAG contains a cycle — topological sort failed")

    return DAGInfo(
        num_ops=num_ops,
        num_tensors=num_tensors,
        adjacency=adjacency,
        reverse_adj=reverse_adj,
        tensor_producer=tensor_producer,
        tensor_consumers=tensor_consumers,
        graph_inputs=graph_inputs,
        graph_outputs=graph_outputs,
        topo_order=topo_order,
    )


def get_subgraph_boundary_tensors(
    op_indices: list[int],
    problem: Problem,
    dag: DAGInfo,
) -> tuple[list[int], list[int], set[int]]:
    """Identify boundary inputs, boundary outputs, and ephemeral tensors for a subgraph.

    Boundary inputs: tensors consumed by the subgraph but NOT produced within it.
    Boundary outputs: tensors produced by the subgraph that are NOT solely consumed within it,
                      OR are graph outputs.
    Ephemeral tensors: produced AND consumed only within the same subgraph (zero cost).

    Args:
        op_indices: Op indices in the subgraph (in topological order).
        problem: The problem spec.
        dag: Pre-computed DAG info.

    Returns:
        (boundary_inputs, boundary_outputs, ephemeral_tensor_ids)
    """
    op_set = set(op_indices)

    produced_within: set[int] = set()
    consumed_within: set[int] = set()

    for op_idx in op_indices:
        op = problem.ops[op_idx]
        for t in op.outputs:
            produced_within.add(t)
        for t in op.inputs:
            consumed_within.add(t)

    # Boundary inputs: consumed within but not produced within
    boundary_inputs = [
        t for t in consumed_within if t not in produced_within
    ]

    # A tensor produced within is ephemeral if ALL its consumers are within the subgraph
    # and it is not a graph output
    ephemeral: set[int] = set()
    boundary_outputs_set: set[int] = set()

    for t in produced_within:
        consumers = dag.tensor_consumers[t]
        all_internal = all(c in op_set for c in consumers)
        is_graph_output = t in dag.graph_outputs

        if all_internal and not is_graph_output and consumers:
            ephemeral.add(t)
        else:
            boundary_outputs_set.add(t)

    boundary_outputs = list(boundary_outputs_set)

    return boundary_inputs, boundary_outputs, ephemeral

File: solution/backend/rust/src/dag.rs
/// DAG utilities: topological sort (Kahn's), adjacency, graph input/output identification.

use std::collections::{HashSet, VecDeque};

use crate::models::Problem;

#[derive(Debug)]
pub struct DagInfo {
    pub num_ops: usize,
    pub num_tensors: usize,
    /// op -> list of successor ops (ops that consume this op's outputs)
    pub successors: Vec<Vec<usize>>,
    /// op -> list of predecessor ops (ops whose outputs feed this op)
    pub predecessors: Vec<Vec<usize>>,
    /// tensor_id -> op that produces it (-1 equivalent = None for graph inputs)
    pub tensor_producer: Vec<Option<usize>>,
    /// tensor_id -> ops that consume it
    pub tensor_consumers: Vec<Vec<usize>>,
    /// tensor indices with no producer (graph inputs, start in slow memory)
    pub graph_inputs: HashSet<usize>,
    /// tensor indices with no consumer (graph outputs, must end in slow memory)
    pub graph_outputs: HashSet<usize>,
    /// topologically sorted op indices
    pub topo_order: Vec<usize>,
}

impl DagInfo {
    pub fn build(problem: &Problem) -> Result<Self, String> {
        let num_ops = problem.ops.len();
        let num_tensors = problem.tensors.len();

        let mut tensor_producer: Vec<Option<usize>> = vec![None; num_tensors];
        let mut tensor_consumers: Vec<Vec<usize>> = vec![vec![]; num_tensors];

        for (op_idx, op) in problem.ops.iter().enumerate() {
            for &out_t in &op.outputs {
                tensor_producer[out_t] = Some(op_idx);
            }
            for &in_t in &op.inputs {
                tensor_consumers[in_t].push(op_idx);
            }
        }

        // Graph inputs: tensors with no producer
        let graph_inputs: HashSet<usize> = (0..num_tensors)
            .filter(|&t| tensor_producer[t].is_none())
            .collect();

        // Graph outputs: tensors with no consumer
        let graph_outputs: HashSet<usize> = (0..num_tensors)
            .filter(|&t| tensor_consumers[t].is_empty())
            .collect();

        // Build op-level adjacency
        let mut successors: Vec<Vec<usize>> = vec![vec![]; num_ops];
        let mut predecessors: Vec<Vec<usize>> = vec![vec![]; num_ops];
        let mut in_degree: Vec<usize> = vec![0; num_ops];

        for (op_idx, op) in problem.ops.iter().enumerate() {
            for &out_t in &op.outputs {
                for &consumer_op in &tensor_consumers[out_t] {
                    successors[op_idx].push(consumer_op);
                    predecessors[consumer_op].push(op_idx);
                    in_degree[consumer_op] += 1;
                }
            }
        }

        // Kahn's algorithm for topological sort
        let mut queue: VecDeque<usize> = (0..num_ops)
            .filter(|&i| in_degree[i] == 0)
            .collect();

        let mut topo_order: Vec<usize> = Vec::with_capacity(num_ops);
        while let Some(op_idx) = queue.pop_front() {
            topo_order.push(op_idx);
            for &succ in &successors[op_idx] {
                in_degree[succ] -= 1;
                if in_degree[succ] == 0 {
                    queue.push_back(succ);
                }
            }
        }

        if topo_order.len() != num_ops {
            return Err("DAG has a cycle".to_string());
        }

        Ok(DagInfo {
            num_ops,
            num_tensors,
            successors,
            predecessors,
            tensor_producer,
            tensor_consumers,
            graph_inputs,
            graph_outputs,
            topo_order,
        })
    }

    /// Given a set of op indices (subgraph), return the boundary input tensor indices:
    /// tensors consumed by ops in the subgraph that are NOT produced within the subgraph.
    pub fn boundary_inputs(
        &self,
        problem: &Problem,
        subgraph_ops: &[usize],
    ) -> Vec<usize> {
        let op_set: HashSet<usize> = subgraph_ops.iter().copied().collect();
        let mut result: HashSet<usize> = HashSet::new();

        for &op_idx in subgraph_ops {
            for &t in &problem.ops[op_idx].inputs {
                // Boundary input if producer is not in the subgraph
                let produced_inside = self.tensor_producer[t]
                    .map(|prod| op_set.contains(&prod))
                    .unwrap_or(false);
                if !produced_inside {
                    result.insert(t);
                }
            }
        }

        let mut v: Vec<usize> = result.into_iter().collect();
        v.sort_unstable();
        v
    }

    /// Given a set of op indices, return boundary output tensor indices:
    /// tensors produced by ops in the subgraph that are either graph outputs
    /// OR consumed by ops OUTSIDE the subgraph.
    pub fn boundary_outputs(
        &self,
        problem: &Problem,
        subgraph_ops: &[usize],
    ) -> Vec<usize> {
        let op_set: HashSet<usize> = subgraph_ops.iter().copied().collect();
        let mut result: HashSet<usize> = HashSet::new();

        for &op_idx in subgraph_ops {
            for &t in &problem.ops[op_idx].outputs {
                // Check if it's a graph output
                if self.graph_outputs.contains(&t) {
                    result.insert(t);
                    continue;
                }
                // Check if any consumer is outside the subgraph
                let has_external_consumer = self.tensor_consumers[t]
                    .iter()
                    .any(|c| !op_set.contains(c));
                if has_external_consumer {
                    result.insert(t);
                }
            }
        }

        let mut v: Vec<usize> = result.into_iter().collect();
        v.sort_unstable();
        v
    }

    /// Returns the set of tensors that are purely ephemeral within a subgraph:
    /// produced AND all consumers are within the same subgraph.
    pub fn ephemeral_tensors(
        &self,
        problem: &Problem,
        subgraph_ops: &[usize],
    ) -> HashSet<usize> {
        let op_set: HashSet<usize> = subgraph_ops.iter().copied().collect();
        let mut result = HashSet::new();

        for &op_idx in subgraph_ops {
            for &t in &problem.ops[op_idx].outputs {
                let all_consumers_inside = self.tensor_consumers[t]
                    .iter()
                    .all(|c| op_set.contains(c));
                let has_consumers = !self.tensor_consumers[t].is_empty();
                if has_consumers && all_consumers_inside && !self.graph_outputs.contains(&t) {
                    result.insert(t);
                }
            }
        }
        result
    }

    /// Determine the output tensor dimension (W_out, H_out) for a subgraph.
    /// For a valid subgraph, all boundary outputs share the same spatial dimensions.
    /// We use the first boundary output tensor's dimensions.
    pub fn output_dimensions(
        &self,
        problem: &Problem,
        subgraph_ops: &[usize],
    ) -> (i64, i64) {
        let boundary_out = self.boundary_outputs(problem, subgraph_ops);
        if boundary_out.is_empty() {
            // Fallback: use output of first op
            let first_op = subgraph_ops[0];
            let out_t = problem.ops[first_op].outputs[0];
            return (problem.tensors[out_t].width, problem.tensors[out_t].height);
        }
        let t = boundary_out[0];
        (problem.tensors[t].width, problem.tensors[t].height)
    }
}
